### PR TITLE
Renew cert fix

### PR DESCRIPTION
**Pull request by mykola-kobets-epam**

### CODE REVIEW

File: iamclient/iamclient.go [hunk 1]
```diff
@@ -37,16 +35,32 @@ import (
...
+	sync.Mutex
+
+	config        *config.Config
+	cryptocontext *cryptutils.CryptoContext
+	insecure      bool
+
 	connection *grpc.ClientConn
-	service    pb.IAMPublicServiceClient
+
+	tlsCertChan      <-chan *pb.CertInfo
```
Comment (al1img) [inline iamclient/iamclient.go:58]: `certChannel`

File: iamclient/iamclient.go [hunk 2]
```diff
@@ -55,11 +69,16 @@ type Client struct {
...
+		IAMPublicServiceClient: grpchelpers.NewIAMPublicServiceClient(iamRequestTimeout),
+
+		config:        config,
+		cryptocontext: cryptocontext,
+		insecure:      insecure,
 
-	if client.connection, client.service, err = client.createConnection(
-		cryptocontext, config.IAMPublicServerURL, insecure); err != nil {
-		return client, err
+		tlsCertChan:      make(<-chan *pb.CertInfo),
```
Comment (al1img) [inline iamclient/iamclient.go:79]: remove make

File: iamclient/iamclient.go [hunk 3]
```diff
@@ -68,93 +87,133 @@ func New(config *config.Config, cryptocontext *cryptutils.CryptoContext, insecur
...
 	}()
 
+	if err = client.openGRPCConnection(); err != nil {
+		return nil, aoserrors.Wrap(err)
+	}
+
+	if !insecure {
+		if ch, err := client.SubscribeCertChanged(config.CertStorage); err != nil {
+			return nil, aoserrors.Wrap(err)
+		} else {
```
Comment (al1img) [inline iamclient/iamclient.go:97]: No else after return

File: iamclient/iamclient.go [hunk 4]
```diff
@@ -68,93 +87,133 @@ func New(config *config.Config, cryptocontext *cryptutils.CryptoContext, insecur
...
+			client.tlsCertChan = ch
```
Comment (al1img) [inline iamclient/iamclient.go:98]: Pass cert channel as parameter instead of private member.

File: iamclient/iamclient.go [hunk 5]
```diff
@@ -68,93 +87,133 @@ func New(config *config.Config, cryptocontext *cryptutils.CryptoContext, insecur
...
+}
+
+func (client *Client) processEvents() {
+	for {
+		select {
+		case <-client.closeChannel:
+			return
+
+		case <-client.tlsCertChan:
+			client.Lock()
```
Comment (al1img) [inline iamclient/iamclient.go:183]: Move lock inside reconnect

File: umclient/umclient.go [hunk 6]
```diff
@@ -345,6 +341,25 @@ func (client *Client) processMessages() (err error) {
...
+			return
+
+		case <-client.certChannel:
+			log.Debug("TLS certificate changed")
+
+			client.closeGRPCConnection()
+
+		case status := <-client.messageHandler.StatusChannel():
+			if err := client.sendStatus(status); err != nil {
+				log.Errorf("Can't send status: %s", aoserrors.Wrap(err))
```
Comment (al1img) [inline umclient/umclient.go:355]: %v

File: umclient/umclient.go [hunk 7]
```diff
@@ -213,74 +201,82 @@ func (client *Client) createConnection(
...
+		for {
+			client.messageHandler.Registered()
 
-		secureOpt = grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig))
-	}
+			if err = client.processGRPCMessages(); err != nil {
+				if errors.Is(err, io.EOF) {
+					log.Debug("Connection is closed")
+				} else {
+					log.Errorf("Connection error: %s", aoserrors.Wrap(err))
```
Comment (al1img) [inline umclient/umclient.go:214]: %v

File: umclient/umclient.go [hunk 8]
```diff
@@ -213,74 +201,82 @@ func (client *Client) createConnection(
...
 
-				return
+	log.Debug("Closing CM connection...")
 
-			case <-time.After(reconnectTimeout):
-				err = client.register()
-			}
+	if client.stream != nil {
+		if err := client.stream.CloseSend(); err != nil {
+			log.WithField("err", err).Error("UM Client failed send close")
```
Comment (al1img) [inline umclient/umclient.go:252]: UM client

File: iamclient/iamclient.go (hunk 1 above)
Comment (mykola-kobets-epam) [reply iamclient/iamclient.go]: done

File: iamclient/iamclient.go (hunk 2 above)
Comment (mykola-kobets-epam) [reply iamclient/iamclient.go]: fixed in aos_common

File: iamclient/iamclient.go (hunk 3 above)
Comment (mykola-kobets-epam) [reply iamclient/iamclient.go]: fixed in aos_common

File: iamclient/iamclient.go (hunk 4 above)
Comment (mykola-kobets-epam) [reply iamclient/iamclient.go]: fixed in aos_common

File: umclient/umclient.go [hunk 9]
```diff
@@ -75,6 +75,7 @@ type Client struct {
 	messageHandler MessageHandler
 	umID           string
 	closeChannel   chan struct{}
+	tlsCertChan    <-chan *iamanager.CertInfo
```
[the remark this answers is not in the document]
Comment (mykola-kobets-epam) [reply umclient/umclient.go]: done

File: umclient/umclient.go (hunk 6 above)
Comment (mykola-kobets-epam) [reply umclient/umclient.go:355]: done

File: umclient/umclient.go (hunk 7 above)
Comment (mykola-kobets-epam) [reply umclient/umclient.go:214]: done

**Review by MykolaSolyanko — APPROVED** [review]

File: updatehandler/updatehandler_test.go [hunk 10]
```diff
@@ -260,6 +260,7 @@ func TestUpdate(t *testing.T) {
 	if handler, err = updatehandler.New(cfg, storage, storage); err != nil {
 		t.Fatalf("Can't create update handler: %s", err)
 	}
+
```
Comment (al1img) [inline updatehandler/updatehandler_test.go:263]: Remove this line: new -> defer close is well known pattern used without empty line

File: updatehandler/updatehandler_test.go [hunk 11]
```diff
@@ -826,6 +827,7 @@ func TestVersionInUpdate(t *testing.T) {
...
```
Comment (al1img) [inline updatehandler/updatehandler_test.go:830]: ditto

**Pull request merged**